h2/src/main/org/h2/mvstore/MVStore.java

```diff
@@ -1761,48 +1772,48 @@ private void compactMoveChunks(Iterable<Chunk> move) {
+            assert iterator.hasNext();
+            long leftmostBlock = iterator.next().block;
+            long originalBlokCount = getAfterLastBlock();
             // we need to ensure that chunks moved within the following loop
             // do not overlap with space just released by chunks moved before them,
-            // hence the need to reserve this area [sourceAreaBegin, originalFileSize)
+            // hence the need to reserve this area [leftmostBlock, originalBlokCount)
             for (Chunk chunk : move) {
-                moveChunk(chunk, sourceAreaBegin, originalFileSize);
+                moveChunk(chunk, leftmostBlock, originalBlokCount);
             }
             // update the metadata (hopefully within the file)
-            store(sourceAreaBegin, originalFileSize);
+            store(leftmostBlock, originalBlokCount);
             sync();
 
             Chunk chunkToMove = lastChunk;
-            long biggestFileSize = getFileLengthInUse();
+            long postEvacuationBlockCount = getAfterLastBlock();
 
             boolean movedToEOF = false;
             // move all chunks, which previously did not fit before reserved area
-            // now we can re-use previously reserved area [sourceAreaBegin, originalFileSize),
-            // but need to reserve [originalFileSize, biggestFileSize)
+            // now we can re-use previously reserved area [leftmostBlock, originalBlokCount),
+            // but need to reserve [originalBlokCount, postEvacuationBlockCount)
             for (Chunk c : move) {
                 if (c.block >= originalBlokCount) {
-                    moveChunk(c, originalFileSize, biggestFileSize);
+                    moveChunk(c, originalBlokCount, postEvacuationBlockCount);
                     assert c.block < originalBlokCount;
                     movedToEOF = true;
                 }
             }
-            if ((movedToEOF || chunkToMove.block >= originalBlokCount) &&
-                    moveChunk(chunkToMove, originalFileSize, biggestFileSize)) {
+            assert postEvacuationBlockCount >= getAfterLastBlock();
+
+            if ((movedToEOF || chunkToMove.block >= originalBlokCount)) {
+                moveChunkInside(chunkToMove, originalBlokCount);
+
                 // store a new chunk with updated metadata (hopefully within a file)
-                store(originalFileSize, biggestFileSize);
+                store(originalBlokCount, postEvacuationBlockCount);
                 sync();
                 // and if last chunk did not fit within a file, since we can use
                 // previously reserved area [originalFileSize, biggestFileSize) now,
                 // try to move it closer to BOF
-                boolean moved = moveChunkInside(chunkToMove, biggestFileSize);
-                if (moveChunkInside(lastChunk, biggestFileSize) || moved) {
-                    store(0, 0);
+                boolean moved = moveChunkInside(chunkToMove, postEvacuationBlockCount);
```

Comment:
If chunkToMove was stored into the right(**[originalBlokCount, postEvacuationBlockCount)**) of file, it can't be moved here. The postEvacuationBlockCount condition is a little too big.